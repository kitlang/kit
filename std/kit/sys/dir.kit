include "dirent.h";
include "unistd.h";
include "sys/time.h";

abstract Directory: dirent {
    public static function read(path: Path): DirectoryReader {
        return DirectoryReader.new(path);
    }

    rules {
        ($this.path) => ($this.d_name as CString) as Path;
    }
}

abstract DirectoryReader: Ptr[DIR] {
    public static function new(path: Path) {
        return opendir(path) as DirectoryReader;
    }

    rules {
        (for $ident in $this {$e}) => {
            var __dir = $this;
            do {
                var __next: Ptr[Directory] = readdir(__dir) as Ptr[Directory];
                if __next == null {
                    break;
                } else {
                    var $ident = __next;
                    {$e}
                }
            } while true;
        }
    }
}

implement Iterable(Ptr[Directory]) for DirectoryReader {
    public function iterator(): Box[Iterator[Ptr[Directory]]] {
        return this;
    }
}

implement Iterator[Ptr[Directory]] for DirectoryReader {
    public function next(): (Box[Iterator[Ptr[Directory]]], Option[Ptr[Directory]]) {
        var dir = readdir(this) as Ptr[Directory];
        if dir != null {
            return (this, Some(dir));
        } else {
            return (this, None);
        }
    }
}
