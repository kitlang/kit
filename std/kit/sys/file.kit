include "stdio.h";
include "sys/stat.h";
include "unistd.h";

// enum SeekFrom {
//     Beginning = SEEK_SET;
//     Current = SEEK_CUR;
//     End = SEEK_END;
// }

enum SeekFrom {
    Beginning;
    Current;
    End;
}

abstract File: Ptr[FILE] {
    public static function read(path: Path): File {
        return fopen(path, "rb") as File;
    }

    public static function write(path: Path): File {
        return fopen(path, "wb") as File;
    }

    public static function append(path: Path): File {
        return fopen(path, "ab") as File;
    }

    public static function update(path: Path): File {
        return fopen(path, "rb+") as File;
    }

    public static function exists(path: Path): Bool {
        return access(path, ${F_OK: Int}) != -1;
    }

    public static function remove(path: Path): Bool {
        return remove(path) == 0;
    }

    public function seek(offset: Int, from: SeekFrom) {
        match from {
            Beginning => fseek(this, offset, ${SEEK_SET: Int});
            Current => fseek(this, offset, ${SEEK_CUR: Int});
            End => fseek(this, offset, ${SEEK_END: Int});
        }
    }

    public function readBytes(buf: Ptr[Void], bytes: Size): Size {
        return fread(buf, 1, bytes, this);
    }

    public function writeBytes(buf: Ptr[Void], bytes: Size): Size {
        return fwrite(buf, 1, bytes, this);
    }

    public function close() {
        fclose(this);
    }

    // public function getSize(): Size {
    //     var buf: stat;
    //     fstat(this as Int, buf);
    //     return buf.st_size;
    // }

    /**
     * Returns the size of this file by seeking to the end of the file. Should
     * only be used on a file opened in Read or Update modes.
     */
    public function getSize(): Size {
        var cur = ftell(this);
        this.seek(0, End);
        var last = ftell(this);
        this.seek(cur, Beginning);
        return last;
    }
}

implement Reader for File {
    public function readBytes(buf: Ptr[Void], bytes: Size): Size {
        return this.readBytes(buf, bytes);
    }

    public function close() {
        this.close();
    }
}

implement Writer for File {
    public function writeBytes(buf: Ptr[Void], bytes: Size): Size {
        return this.writeBytes(buf, bytes);
    }

    public function close() {
        this.close();
    }
}
