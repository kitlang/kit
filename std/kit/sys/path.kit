#[promote] abstract Path: CString {
    // FIXME
    public static var PATH_SEP: Char = c'/';

    rules {
        // (${p1: Path} / ${p2: Path}) => $p1.join($p2);
    }

    public function isAbsolute(): Bool {
        return this.length > 0 && this[0] == Self.PATH_SEP;
    }

    public function join(allocator: Box[Allocator], other: Path): Path {
        if other.isAbsolute() {
            return other;
        } else {
            var l1 = this.length;
            var l2 = other.length;
            var addSep = this[l1 - 1] != Self.PATH_SEP;
            var joined: CString = allocator.alloc(this.length + other.length + (if addSep then 2 else 1));
            strncpy(joined, this, l1);
            if addSep {
                joined[l1] = Self.PATH_SEP;
                strcpy(joined + l1 + 1, other);
            } else {
                strcpy(joined + l1, other);
            }
            return joined;
        }
    }
}
