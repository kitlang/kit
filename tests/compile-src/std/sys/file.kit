import kit.sys.file;

function main() {
    var fWrite: File = File.write("test_file");
    var writer: Box[Writer] = fWrite;
    writer.write("hello");
    fWrite.close();

    if !File.exists("test_file") {
        printf("test_file should exist\n");
    }

    var fWrite2: File = File.write("test_file");
    fWrite2.writeBytes("hi ", 3);
    fWrite2.close();

    var fAppend: File = File.append("test_file");
    fAppend.writeBytes("there", 6);
    fAppend.close();

    var fUpdate: File = File.update("test_file");
    fUpdate.writeBytes("hey", 3);
    printf("size after update: %zu\n", fUpdate.getSize());
    fUpdate.close();

    var fRead: File = File.read("test_file");
    var reader: Box[Reader] = fRead;
    var s: Ptr[Void] = malloc(9);
    reader.readBytes(s, 8);
    (s as CString)[8] = 0;
    printf("%s\n", s as CString);
    printf("final size: %zu\n", fRead.getSize());
    fRead.close();

    var removed = File.remove("test_file");
    if !removed {
        printf("test_file should have been removed\n");
    }
    if File.exists("test_file") {
        printf("test_file should have been removed\n");
    }
}
