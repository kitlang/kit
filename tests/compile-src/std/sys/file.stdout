size after update: 9
heythere
final size: 9
